src/producer/TransactionMQProducer.cpp

```diff
@@ -0,0 +1,185 @@
+      localTransactionState = LocalTransactionState::ROLLBACK_MESSAGE;
+      break;
+    default:
+      break;
+  }
+
+  try {
+    endTransaction(sendResult, localTransactionState);
+  } catch (MQException& e) {
+    LOG_WARN("endTransaction exception:%s", e.what());
+  }
+
+  TransactionSendResult transactionSendResult(sendResult.getSendStatus(), sendResult.getMsgId(),
+                                              sendResult.getOffsetMsgId(), sendResult.getMessageQueue(),
+                                              sendResult.getQueueOffset());
+  transactionSendResult.setTransactionId(msg.getTransactionId());
+  transactionSendResult.setLocalTransactionState(localTransactionState);
+  return transactionSendResult;
+}
+
+void TransactionMQProducer::endTransaction(SendResult& sendResult, LocalTransactionState& localTransactionState) {
+  MQMessageId id;
+  if (sendResult.getOffsetMsgId() != "") {
+    id = MQDecoder::decodeMessageId(sendResult.getOffsetMsgId());
+  } else {
+    id = MQDecoder::decodeMessageId(sendResult.getMsgId());
+  }
+  string transId = sendResult.getTransactionId();
+
+  int commitOrRollback = 0;
+  switch (localTransactionState) {
+    case COMMIT_MESSAGE:
+      commitOrRollback = MessageSysFlag::TransactionCommitType;
+      break;
+    case ROLLBACK_MESSAGE:
+      commitOrRollback = MessageSysFlag::TransactionRollbackType;
+      break;
+    case UNKNOW:
+      commitOrRollback = MessageSysFlag::TransactionNotType;
+      break;
+    default:
+      commitOrRollback = MessageSysFlag::TransactionNotType;
+      break;
+  }
+
+  bool fromTransCheck = false;
+  EndTransactionRequestHeader* requestHeader =
+      new EndTransactionRequestHeader(getGroupName(), sendResult.getQueueOffset(), id.getOffset(), commitOrRollback,
+                                      fromTransCheck, sendResult.getMsgId(), transId);
+  LOG_DEBUG("endTransaction: msg:%s", requestHeader->toString().data());
+  getFactory()->endTransactionOneway(sendResult.getMessageQueue(), requestHeader, getSessionCredentials());
+}
+
+void TransactionMQProducer::checkTransactionState(const std::string& addr, const MQMessageExt& message,
+                                                  long m_tranStateTableOffset,
+                                                  long m_commitLogOffset,
+                                                  std::string m_msgId,
+                                                  std::string m_transactionId,
+                                                  std::string m_offsetMsgId) {
+  LocalTransactionState localTransactionState = m_transactionListener->checkLocalTransaction(message);
+
+  EndTransactionRequestHeader* endHeader = new EndTransactionRequestHeader();
+  endHeader->m_commitLogOffset = m_commitLogOffset;
+  endHeader->m_producerGroup = getGroupName();
+  endHeader->m_tranStateTableOffset = m_tranStateTableOffset;
+  endHeader->m_fromTransactionCheck = true;
+
+  string uniqueKey = m_transactionId;
```

Comment:
why not get uniqueKey from PROPERTY_UNIQ_CLIENT_MESSAGE_ID_KEYIDX property of message.

Reply:
same to java client